src/Database/MemoryCache.php

```diff
@@ -0,0 +1,129 @@
+<?php namespace October\Rain\Database;
+
+use October\Rain\Support\Traits\Singleton;
+
+/**
+ * Query memory cache class.
+ *
+ * Stores query results in memory to avoid running duplicate queries
+ *
+ * @package october\database
+ * @author Alexey Bobkov, Samuel Georges
+ */
+class MemoryCache
+{
+    use Singleton;
+
+    /**
+     * Cached results.
+     *
+     * @var array
+     */
+    protected $cache = [];
+
+    /**
+     * The mapper between hashed keys and table names.
+     *
+     * @var array
+     */
+    protected $tableMap = [];
+
+    /**
+     * Check if the memory cache is enabled.
+     *
+     * @return bool
+     */
+    public function enabled()
+    {
+        return true; // TODO: Add a config option
```

Comment:
I left this config option open, since I'm not sure where is the best place to put it.